### PR TITLE
Refactor unchecked std::array subscript operator

### LINKED ISSUE #15618: Ensure that std::array access is within bounds during compile-time
There are a couple cases in our codebase where arrays are accessed with indices known during compile time (hardcoded or `constexpr` values). Using [`std::get<...>(arr)`](https://en.cppreference.com/w/cpp/container/array/get) instead of [`arr[...]`](https://en.cppreference.com/w/cpp/container/array/operator_at) ensures that the indices are within bounds, or it'll simply fail to compile.

```cpp
void test(std::array<int, 10>& arr)
{
    arr[10] = 1;           // Out of bounds, but no error
    arr.at(11) = 2;        // Out of bounds, but no error
    std::get<12>(arr) = 3; // Fails to compile
}
```

A few examples in our codebase are:

https://github.com/OpenRCT2/OpenRCT2/blob/557d4119df61856c3027ec9909858633b8c45367/src/openrct2/world/Footpath.cpp#L1887

https://github.com/OpenRCT2/OpenRCT2/blob/e06f9c3ee70f492b7403d31a5038e34934bc8ffd/src/openrct2/actions/TrackPlaceAction.cpp#L325

https://github.com/OpenRCT2/OpenRCT2/blob/e06f9c3ee70f492b7403d31a5038e34934bc8ffd/src/openrct2/core/Random.hpp#L148-L149

### PR DESCRIPTION
**Pull request by peterschussheim**
Closes #15618

### CODE REVIEW

**Review by Broxzier — APPROVED** [review]

File: src/openrct2/world/Footpath.cpp [hunk 1]
```diff
@@ -1929,12 +1936,14 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
...
 
-        if ((pathConnections & FOOTPATH_CONNECTION_NW) && pathList[3] != nullptr && !pathList[3]->AsPath()->IsWide())
+        if ((pathConnections & FOOTPATH_CONNECTION_NW) && std::get<3>(pathList) != nullptr
+            && !std::get<3>(pathList)->AsPath()->IsWide())
         {
             constexpr uint8_t edgeMask1 = EDGE_SE | EDGE_SW;
-            if ((pathConnections & FOOTPATH_CONNECTION_NE) && pathList[7] != nullptr && !pathList[7]->AsPath()->IsWide()
-                && (pathList[7]->AsPath()->GetEdges() & edgeMask1) == edgeMask1 && pathList[0] != nullptr
-                && !pathList[0]->AsPath()->IsWide())
+            if ((pathConnections & FOOTPATH_CONNECTION_NE) && std::get<7>(pathList) != nullptr
```
Comment (Gymnasiast) [inline src/openrct2/world/Footpath.cpp:1943]: Please put `std::get<7>(pathList)` into a `const auto&` variable.

File: src/openrct2/world/Footpath.cpp [hunk 2]
```diff
@@ -1945,8 +1954,9 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
              * in combination with reset tiles.
              * Short circuit the logic appropriately. */
             constexpr uint8_t edgeMask2 = EDGE_NE | EDGE_SE;
-            if ((pathConnections & FOOTPATH_CONNECTION_SW) && pathList[6] != nullptr && !(pathList[6])->AsPath()->IsWide()
-                && (pathList[6]->AsPath()->GetEdges() & edgeMask2) == edgeMask2 && pathList[2] != nullptr)
+            if ((pathConnections & FOOTPATH_CONNECTION_SW) && std::get<6>(pathList) != nullptr
```
Comment (Gymnasiast) [inline src/openrct2/world/Footpath.cpp:1957]: See previous.

File: src/openrct2/world/Footpath.cpp (hunk 1 above)
Comment (peterschussheim) [reply src/openrct2/world/Footpath.cpp:1943]: There are a few other examples where a particular path element is accessed 2 or more times, is there a preference/rule on when to extract these usages into a variable?  I don't want to make this more complicated however consistency is important.

File: src/openrct2/world/Footpath.cpp [hunk 3]
```diff
@@ -1891,7 +1891,8 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
         if (tileElement->AsPath()->GetEdges() & EDGE_NW)
         {
             pathConnections |= FOOTPATH_CONNECTION_NW;
-            if (std::get<3>(pathList) != nullptr && std::get<3>(pathList)->AsPath()->IsWide())
+            const auto& path_element = std::get<3>(pathList);
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1894]:
```suggestion
            const auto* pathElement = std::get<3>(pathList);
```
Use auto* when its a pointer and auto& when its a reference and just auto when you intent to do a copy. Also our coding style uses ```camelCase``` for variables.

File: src/openrct2/world/Footpath.cpp [hunk 4]
```diff
@@ -1900,7 +1901,8 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
         if (tileElement->AsPath()->GetEdges() & EDGE_NE)
         {
             pathConnections |= FOOTPATH_CONNECTION_NE;
-            if (std::get<0>(pathList) != nullptr && std::get<0>(pathList)->AsPath()->IsWide())
+            const auto& path_element = std::get<0>(pathList);
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1904]: Same here

File: src/openrct2/world/Footpath.cpp [hunk 5]
```diff
@@ -1940,9 +1942,10 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
             && !std::get<3>(pathList)->AsPath()->IsWide())
         {
             constexpr uint8_t edgeMask1 = EDGE_SE | EDGE_SW;
-            if ((pathConnections & FOOTPATH_CONNECTION_NE) && std::get<7>(pathList) != nullptr
-                && !std::get<7>(pathList)->AsPath()->IsWide()
-                && (std::get<7>(pathList)->AsPath()->GetEdges() & edgeMask1) == edgeMask1 && std::get<0>(pathList) != nullptr
+            const auto& path_element1 = std::get<7>(pathList);
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1945]:
```suggestion
            const auto* pathElement7 = std::get<7>(pathList);
```
Makes it more clear that its the 7th element in the list.

File: src/openrct2/world/Footpath.cpp [hunk 6]
```diff
@@ -1940,9 +1942,10 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
...
+            if ((pathConnections & FOOTPATH_CONNECTION_NE) && path_element1 != nullptr
+                && !path_element1->AsPath()->IsWide()
+                && (path_element1->AsPath()->GetEdges() & edgeMask1) == edgeMask1 && std::get<0>(pathList) != nullptr
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1947]: Also extract a variable for <0>

File: src/openrct2/world/Footpath.cpp [hunk 7]
```diff
@@ -1954,9 +1957,10 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
              * in combination with reset tiles.
              * Short circuit the logic appropriately. */
             constexpr uint8_t edgeMask2 = EDGE_NE | EDGE_SE;
-            if ((pathConnections & FOOTPATH_CONNECTION_SW) && std::get<6>(pathList) != nullptr
-                && !(std::get<6>(pathList))->AsPath()->IsWide()
-                && (std::get<6>(pathList)->AsPath()->GetEdges() & edgeMask2) == edgeMask2 && std::get<2>(pathList) != nullptr)
+            const auto& path_element2 = std::get<6>(pathList);
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1959]:
```suggestion
            const auto* pathElement6 = std::get<6>(pathList);
```

File: src/openrct2/world/Footpath.cpp [hunk 8]
```diff
@@ -1954,9 +1957,10 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
...
+            if ((pathConnections & FOOTPATH_CONNECTION_SW) && path_element2 != nullptr
+                && !(path_element2)->AsPath()->IsWide()
+                && (path_element2->AsPath()->GetEdges() & edgeMask2) == edgeMask2 && std::get<2>(pathList) != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1961]: Also extract variable for <2>

File: src/openrct2/world/Footpath.cpp [hunk 9]
```diff
@@ -1971,8 +1975,9 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
         if ((pathConnections & FOOTPATH_CONNECTION_SE) && std::get<1>(pathList) != nullptr)
         {
             constexpr uint8_t edgeMask1 = EDGE_SW | EDGE_NW;
-            if ((pathConnections & FOOTPATH_CONNECTION_NE) && (std::get<4>(pathList) != nullptr)
-                && (std::get<4>(pathList)->AsPath()->GetEdges() & edgeMask1) == edgeMask1 && std::get<0>(pathList) != nullptr
+            const auto& path_element1 = std::get<4>(pathList);
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1976]:
```suggestion
            const auto* pathElement4 = std::get<4>(pathList);
```

File: src/openrct2/world/Footpath.cpp [hunk 10]
```diff
@@ -1971,8 +1975,9 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
...
+            if ((pathConnections & FOOTPATH_CONNECTION_NE) && (path_element1 != nullptr)
+                && (path_element1->AsPath()->GetEdges() & edgeMask1) == edgeMask1 && std::get<0>(pathList) != nullptr
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1978]: Extract also <0>

File: src/openrct2/world/Footpath.cpp [hunk 11]
```diff
@@ -1985,8 +1990,9 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
              * in combination with reset tiles.
              * Short circuit the logic appropriately. */
             constexpr uint8_t edgeMask2 = EDGE_NE | EDGE_NW;
-            if ((pathConnections & FOOTPATH_CONNECTION_SW) && std::get<5>(pathList) != nullptr
-                && (std::get<5>(pathList)->AsPath()->GetEdges() & edgeMask2) == edgeMask2 && std::get<2>(pathList) != nullptr)
+            const auto& path_element2 = std::get<5>(pathList);
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1991]:
```suggestion
            const auto* pathElement5 = std::get<5>(pathList);
```

File: src/openrct2/world/Footpath.cpp [hunk 12]
```diff
@@ -1985,8 +1990,9 @@ void footpath_update_path_wide_flags(const CoordsXY& footpathPos)
...
+            if ((pathConnections & FOOTPATH_CONNECTION_SW) && path_element2 != nullptr
+                && (path_element2->AsPath()->GetEdges() & edgeMask2) == edgeMask2 && std::get<2>(pathList) != nullptr)
```
Comment (ZehMatt) [inline src/openrct2/world/Footpath.cpp:1993]: And <2>

**Review by ZehMatt — CHANGES_REQUESTED** [review]

File: src/openrct2/world/Footpath.cpp (hunk 3 above)
Comment (peterschussheim) [reply src/openrct2/world/Footpath.cpp:1894]: Got it, thank you for that feedback

**Review by Gymnasiast — APPROVED** [review]

**Comment by Broxzier** [comment]
@peterschussheim this PR is still a draft. Do you intend to do more work, or can this be merged? If you want to do more, it may be best to do that in a new PR to keep conflicts to a minimum.

**Comment by peterschussheim** [comment]
Thanks for the reminder, I will wrap this PR up today :)

**Review by ZehMatt — APPROVED** [review]

**Pull request merged**